### ISSUE TITLE
MirthCorp: DB Writer keeps previous URL in cache and retries

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3575]
Hi Nick,
This is the test that we did. Test environment was an appliance used for training. We create a DB, and then a table. Then when we create a new channel and then as destination choose DB Writer in non JavaScript mode, if you input the wrong URL and click on Insert and Get Tables, can not connect. Which is the right behavior, however if you correct the URL to localhost and then try the same steps, still can not connect to DB as it tries the old entry in cache

Cheers,

Sean

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3687
Reporter: wilsont
Created: 2015-04-08T12:45:01.000-0700

**Comment by rbeckman-nextgen** [comment]
I have been unable to replicate this issue using the above steps. At this point, it sounds like it might have been a configuration problem rather than a cache one.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-04-13T15:42:54.000-0700

**Comment by rbeckman-nextgen** [comment]
So, Sean came over today and we seemingly reproduced what he thinks to be the problem. This seems to be a database config issue not allowing the connection to process via the DNS. After some time, an error is thrown saying that the database metadata could not be retrieved.

Not sure if we want to make the timeout configurable or do anything at all.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-04-14T17:46:36.000-0700

**Comment by rbeckman-nextgen** [comment]
In his example, shouldn't localhost connect just fine though? Why is an error thrown at all? The way I understand it, if he tries the incorrect domain name, it fails, which is correct. However after failing if he switches the domain name back to the correct one (in his case localhost), then it still fails, when it _should_ succeed. Is that correct [~seanm]?

Imported Comment. Original Details:
Author: narupley
Created: 2015-04-14T18:31:15.000-0700

**Comment by rbeckman-nextgen** [comment]
Hi Nick,
Yes, after having the right URL it should work. I showed Jason yesterday how to reproduce the bug using a Training Appliance and the scenario that happens in the exam. 

Imported Comment. Original Details:
Author: seanm
Created: 2015-04-15T08:21:43.000-0700

**Comment by rbeckman-nextgen** [comment]
Localhost is working just fine.

What appears to be happening is that the initial url seemingly isn't working as the connection is hanging during establishing a connection. It seems to be waiting the full login timeout duration of 30 seconds. During this time, we were closing the DatabaseMetadataDialog and changing the url while the previous connection is still stuck making the connection and attempting to get tables with the corrected url pointed to localhost. In my IDE, the background thread is getting cancelled but the dialog/admin is still unresponsive until the timeout duration is exceeded.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-04-15T10:09:33.000-0700

**Comment by rbeckman-nextgen** [comment]
But when the incorrect domain name is used, don't you still get the "could not retrieve" error dialog? After getting that error dialog, why is anything still happening in a background thread?

Imported Comment. Original Details:
Author: narupley
Created: 2015-04-15T11:48:18.000-0700

**Comment by rbeckman-nextgen** [comment]
When the incorrect domain name is used and the thread finishes to completion, yes, that error dialog appears.

When Sean came over, we were closing out the 'Insert' dialog and changing the url before the error dialog ever appeared. So, when we tried with the corrected url, it appeared as if it wasn't working when in actuality, it was still timing out from the last connection it seems. I tried it again after he left and waited to see what happened when I didn't close the dialog and I got the error message that you see in the screenshot. I was able to then correctly update the database url and it worked without incident. Looking at the code though, it seems this working thread should have been cancelled at the time a new request was made to get the tables.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-04-15T12:32:24.000-0700

**Comment by rbeckman-nextgen** [comment]
Makes sense... so this is still a valid bug then right?

This is actually a common problem with connector services or any request that does something in the background and then updates the UI or shows a dialog in the EDT after it finishes. You can do similar things with the other services (like Get Operations in Web Service Sender) that don't make sense per se.

Because this is a more general problem and not some kind of regression bug introduced in 3.2, I'll take it out of 3.2.2, unless you have any other thoughts or concerns...

Imported Comment. Original Details:
Author: narupley
Created: 2015-04-15T13:11:50.000-0700

**Comment by rbeckman-nextgen** [comment]
For now, we'll create a separate method in Client, invokeConnectorServiceAsync, basically the same thing we did for invokePluginMethod. Then the DatabaseMetadataDialog will call the new async method. Because it is correctly closing its workers when the dialog is closed, it's okay to allow this method to go on an asynchronous connection. Not all places in the Administrator yet do this correctly, so for the more generic problem of migrating to the new method (only where appropriate) I've created a separate issue: MIRTH-3752

Imported Comment. Original Details:
Author: narupley
Created: 2015-07-24T14:34:36.000-0700

**Comment by rbeckman-nextgen** [comment]
Pulling tables for the DB Writer no longer does a synchronous call which was thought to cause the above problem. The threadworker is now correctly cancelled.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-07-24T14:49:23.000-0700

**Comment by rbeckman-nextgen** [comment]
**Server(s) and URLs Used:** Run Mirth Connect on local OSX Yosemite 10.10
**Version(s)/Build(s) to reproduce failure:** mirthconnect-3.1.1.7461.b23-unix
**Version(s)/Build(s) to verify fixes:** mirthconnect-3.3.0.7719.b1682-unix
**How Tested (list tests run):**
* Prior to testing, we need to have a working Postges database
** On minh-vwin7.hq.mirthcorp.com, setup postgres
** Create database name = mirthtraining
** Create one or two tables in the database 
* Start Mirth Connect
* Create new channel 
Source = any, in this test, I used the default Channel Reader
Destination = Database Writer
**  Driver = PostgreSQL
**  Click on Insert URL template to fill out the URL
**  Put in the username & password for the database
* Update URL with the correct hostname & database name from Step 1, then click on Insert > Get Tables 
=> Tables should be are listed promptly, then select Cancel to close.
For this test, I used
` jdbc:postgresql://minh-vwin7.hq.mirthcorp.com:5432/mirthtraining `
* Now modify URL with a hostname is currently not available, such as a shutdown VM and click on Insert > Get Tables 
=> no table listed, should see “Retrieve tables running” on bottom bar, select Cancel to close
For this test, I used 
` jdbc:postgresql://minh-vwin8.hq.mirthcorp.com:5432/mirthtraining `
* Repeat the process with the correct hostname, record behavior

**Verified Fixed:**
In new build, the updated URL got responded right away whether the previous one has timed out or not  
In previous build, the updated URL got no response until after the previous request timed out.
 
**Additional Information**
The challenge to recreate this issue was how long it takes for the wrong connection to timeout ( Error Popup once request completes). I was getting the Error "Could not retrieve database..." almost right away when using wrong hostname or wrong DNS . Whereas it could take a minute or two when using an unavailable host. I suspect this could be potential issue for environment with slow DNS lookup.

Imported Comment. Original Details:
Author: minht
Created: 2015-08-20T17:18:32.000-0700
